pkg/controller/gcp/database/operations.go

```diff
@@ -0,0 +1,252 @@
+	sqladmin "google.golang.org/api/sqladmin/v1beta4"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	corev1alpha1 "github.com/crossplaneio/crossplane/pkg/apis/core/v1alpha1"
+	"github.com/crossplaneio/crossplane/pkg/apis/gcp/database/v1alpha1"
+	"github.com/crossplaneio/crossplane/pkg/clients/gcp/cloudsql"
+	"github.com/crossplaneio/crossplane/pkg/meta"
+	"github.com/crossplaneio/crossplane/pkg/util"
+)
+
+type localOperations interface {
+	// Bucket object managedOperations
+	addFinalizer(context.Context) error
+	isReclaimDelete() bool
+	isInstanceReady() bool
+	needUpdate(*sqladmin.DatabaseInstance) bool
+	removeFinalizer(context.Context) error
+
+	// Controller-runtime managedOperations
+	updateObject(ctx context.Context) error
+	updateInstanceStatus(context.Context, *sqladmin.DatabaseInstance) error
+	updateReconcileStatus(context.Context, error) error
+	updateConnectionSecret(ctx context.Context) (*corev1.Secret, error)
+}
+
+type localHandler struct {
+	*v1alpha1.CloudsqlInstance
+	client client.Client
+}
+
+var _ localOperations = &localHandler{}
+
+func newLocalHandler(instance *v1alpha1.CloudsqlInstance, kube client.Client) *localHandler {
+	return &localHandler{
+		CloudsqlInstance: instance,
+		client:           kube,
+	}
+}
+
+//
+// Crossplane GCP Bucket object managedOperations
+//
+func (h *localHandler) addFinalizer(ctx context.Context) error {
+	meta.AddFinalizer(h, finalizer)
+	return h.updateObject(ctx)
+}
+
+func (h *localHandler) removeFinalizer(ctx context.Context) error {
+	meta.RemoveFinalizer(h, finalizer)
+	return h.updateObject(ctx)
+}
+
+func (h *localHandler) isInstanceReady() bool {
+	return h.IsAvailable()
+}
+
+func (h *localHandler) isReclaimDelete() bool {
+	return h.Spec.ReclaimPolicy == corev1alpha1.ReclaimDelete
+}
+
+func (h *localHandler) needUpdate(actual *sqladmin.DatabaseInstance) bool {
+	// TODO: update functionality is not supported for this instance.
+	//   In order to add this support we need to refactor DatabaseInstanceType
+	return false
+
+	// TODO: when we ready to support update, delete 4 lines above and uncomment the lines below
+	//desired := h.DatabaseInstance(h.GetResourceName())
+	//if diff := cmp.Diff(actual, desired); diff != "" {
+	//	log.V(logging.Debug).Info(diff)
+	//	return false
+	//}
+	//return true
+}
+
+func (h *localHandler) updateObject(ctx context.Context) error {
+	return h.client.Update(ctx, h.CloudsqlInstance)
+}
+
+func (h *localHandler) updateInstanceStatus(ctx context.Context, inst *sqladmin.DatabaseInstance) error {
+	h.SetStatus(inst)
+	return h.client.Status().Update(ctx, h.CloudsqlInstance)
+}
+
+func (h *localHandler) updateReconcileStatus(ctx context.Context, err error) error {
+	if err == nil {
+		h.Status.SetConditions(corev1alpha1.ReconcileSuccess())
+	} else {
+		h.Status.SetConditions(corev1alpha1.ReconcileError(err))
+	}
+	return h.client.Status().Update(ctx, h.CloudsqlInstance)
+}
```

Comment:
I personally feel the controller code becomes harder to follow as we bury things like this (setting conditioned status)  into helper functions and layers of abstraction. This relates to our discussion on Friday in which I mentioned that in accepting a slightly higher level of cyclomatic complexity the code felt like it became easier to follow (and perhaps even easier to test).

Reply:
I can see what you mean. Not sure if it is a good or bad thing.